File: epi_judge_cpp/majority_element.cc
#include <string>
#include <vector>

#include "test_framework/generic_test.h"
using std::string;
using std::vector;

string MajoritySearch(vector<string>::const_iterator stream_begin,
                      const vector<string>::const_iterator stream_end) {
    int candidate_count = 0;
    string candidate;
    vector<string>::const_iterator iter = stream_begin;
    
    while (iter != stream_end) {
        if (candidate_count == 0){
            candidate = *iter;
            candidate_count = 1;
        }
        else if (candidate == *iter) {
            candidate_count++;
        }
        else {
            candidate_count--;
        }
    
        iter++;
    }

    return candidate;
}
string MajoritySearchWrapper(const vector<string>& stream) {
  return MajoritySearch(cbegin(stream), cend(stream));
}

int main(int argc, char* argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"stream"};
  return GenericTestMain(args, "majority_element.cc", "majority_element.tsv",
                         &MajoritySearchWrapper, DefaultComparator{},
                         param_names);
}

File: epi_judge_python_solutions/majority_element.py
def majority_search(stream):
    candidate, candidate_count = None, 0
    for it in stream:
        if candidate_count == 0:
            candidate, candidate_count = it, candidate_count + 1
        elif candidate == it:
            candidate_count += 1
        else:
            candidate_count -= 1
    return candidate


def majority_search_wrapper(stream):
    return majority_search(iter(stream))


from sys import exit

from test_framework import generic_test, test_utils

if __name__ == '__main__':
    exit(
        generic_test.generic_test_main('majority_element.tsv',
                                       majority_search_wrapper))
